feature_engine/encoding/similarity_encoder.py

```diff
@@ -0,0 +1,298 @@
+from typing import List, Optional, Union
+
+import difflib
+import numpy as np
+import pandas as pd
+from sklearn.utils.validation import check_is_fitted
+
+from feature_engine._docstrings.fit_attributes import (
+    _feature_names_in_docstring,
+    _n_features_in_docstring,
+    _variables_attribute_docstring,
+)
+from feature_engine._docstrings.init_parameters import (
+    _ignore_format_docstring,
+    _variables_categorical_docstring,
+)
+from feature_engine._docstrings.methods import _fit_transform_docstring
+from feature_engine._docstrings.substitute import Substitution
+from feature_engine.dataframe_checks import _check_contains_na, check_X
+from feature_engine.encoding.base_encoder import (
+    CategoricalInitMixin,
+    CategoricalMethodsMixin,
+)
+
+
+def _gpm_fast(x1: str, x2: str) -> float:
+    return difflib.SequenceMatcher(None, x1, x2).quick_ratio()
+
+
+_gpm_fast_vec = np.vectorize(_gpm_fast)
+
+
+@Substitution(
+    ignore_format=_ignore_format_docstring,
+    variables=_variables_categorical_docstring,
+    variables_=_variables_attribute_docstring,
+    feature_names_in_=_feature_names_in_docstring,
+    n_features_in_=_n_features_in_docstring,
+    fit_transform=_fit_transform_docstring,
+)
+class StringSimilarityEncoder(CategoricalInitMixin, CategoricalMethodsMixin):
+    """
+    The StringSimilarityEncoder() replaces categorical variables by a set of
+    float variables representing similarity between unique categories in the variable.
+    This new variables will have values in range between 0 and 1, where 0 is the least
+    similar and 1 is the exact match.
+    This encoding is an alternative to OneHotEncoder in the case of poorly
```

Comment:
Should we add information regarding how the similarity is calculated? I'd be curios if I was a user. Is it too long to explain?

Reply:
Ye, it's not very straightforward, but we can sen user to difflib.SequenceMatcher class user doc

Reply:
I think we can borrow these 3 sentences from the difflib library:

Return a measure of the sequences’ similarity as a float in the range [0, 1].

Where T is the total number of elements in both sequences, and M is the number of matches, this is 2.0*M / T. Note that this is 1.0 if the sequences are identical, and 0.0 if they have nothing in common.

And mention that this transformer uses difflib.SequenceMatcher under the hood, for the most experienced user.